### ISSUE TITLE
EventProcessorClient failing to Start when Fully Qualified Namespace is malformed; Error Handler not invoked.

### ISSUE THREAD

**Issue by joebeernink** [Azure/azure-sdk-for-net#10576]
Have the following method for Starting the EventClientProcessor. The Client never gets to running state, no matter how long we wait. Must be missing something simple, but there are no errors thrown. Any ideas?

  ```
 public override async Task StartAsync(CancellationToken cancellationToken)
        {
            await base.StartAsync(cancellationToken).ConfigureAwait(false);
            logger.LogInformation("ApimLogToGeneva starting");

            var eventHubConfiguration = dynamicConfigurationProvider.Get<EventHubConfiguration>();
            var apimLogToGenevaConfiguration = dynamicConfigurationProvider.Get<ApimLogToGenevaConfiguration>();

            async Task ProcessEventHandler(ProcessEventArgs eventArgs)
            {
                try
                {
                    logger.LogInformation("Event Received on Partition: {partitionId}", eventArgs.Partition.PartitionId);

                    // Perform the application-specific processing for an event
                    await apimEventHandler.ProcessEventAsync(eventArgs.Partition.PartitionId, eventArgs.Data, cancellationToken).ConfigureAwait(false);
                }
                catch (Exception ex)
                {
                    // Handle the exception from handler code
                    logger.LogError(ex, "Exception occurred while processing the event handler for PartitionId {partitionId}", eventArgs.Partition.PartitionId);
                }
            }

            async Task ProcessErrorHandler(ProcessErrorEventArgs eventArgs)
            {
                try
                {
                    logger.LogInformation("Error received on Partition: {partitionId}", eventArgs.PartitionId);

                    // Perform the application-specific processing for an error
                    await apimEventHandler.ProcessErrorAsync(eventArgs.PartitionId, eventArgs.Exception).ConfigureAwait(false);
                }
                catch (Exception ex)
                {
                    // Handle the exception from handler code
                    logger.LogError(ex, "Exception occurred while processing the error handler for PartitionId {partitionId}", eventArgs.PartitionId);
                }
            }

            try
            {
                eventProcessorClient = tokenProvider.GetEventProcessorClient();

                logger.LogInformation("Got Token Credential for Event Hub Namespace: {namespace}, Event Hub Name: {name}, Consumer Group: {consumerGroup}", eventHubConfiguration.Namespace, eventHubConfiguration.Name, eventHubConfiguration.ConsumerGroup);

                logger.LogInformation("Adding Handlers");
                eventProcessorClient.ProcessEventAsync += ProcessEventHandler;
                eventProcessorClient.ProcessErrorAsync += ProcessErrorHandler;

                logger.LogInformation("Starting Processor.");

                await eventProcessorClient.StartProcessingAsync().ConfigureAwait(false);

                while (!eventProcessorClient.IsRunning && !cancellationToken.IsCancellationRequested)
                {
                    logger.LogInformation("Client status. IsRunning: {isRunning}", eventProcessorClient.IsRunning);
                    await Task.Delay(TimeSpan.FromSeconds(10)).ConfigureAwait(false);
                }

                logger.LogInformation("Client status. IsRunning: {isRunning}", eventProcessorClient.IsRunning);

                while (!cancellationToken.IsCancellationRequested)
                {
                    await Task.Delay(TimeSpan.FromSeconds(1)).ConfigureAwait(false);
                }

                logger.LogInformation("Shutting down APIM Worker Service");
                await eventProcessorClient.StopProcessingAsync().ConfigureAwait(false);
            }
            catch (Exception ex)
            {
                logger.LogError(ex, "Exception occurred while setting up the eventProcessorClient");
            }
            finally
            {
                // To prevent leaks, the handlers should be removed when processing is complete
                eventProcessorClient.ProcessEventAsync -= ProcessEventHandler;
                eventProcessorClient.ProcessErrorAsync -= ProcessErrorHandler;
            }
        }
```

**Comment by AlexGhiondea** [comment]
@joebeernink could you provide us with more information? Below is bug issue template:

**Describe the bug**
Please provide the description of issue you're seeing.

**Expected behavior**
What is the expected behavior?

**Actual behavior (include Exception or Stack Trace)**
What is the actual behavior?

**To Reproduce**
Steps to reproduce the behavior (include a code snippet, screenshot, or any additional information that might help us reproduce the issue)

1.
2.

**Environment:**
 - Name and version of the Library package used: [e.g. Azure.Storage.Blobs 12.2.0] 
 - Hosting platform or OS and .NET runtime version (`dotnet --info` output for .NET Core projects): [e.g. Azure AppService or Windows 10 .NET Framework 4.8]
 - IDE and version : [e.g. Visual Studio 16.3]

**Comment by jsquire** [comment]
Hi @joebeernink.  We're sorry that you're running into trouble and appreciate you reaching out to let us know.   I don't see anything obvious in your code at a glance.  That said, I'd like confirm that when you say that you don't see it "get into a running state" you mean that you don't see the `IsRunning` property set and are spending time in this loop:

```csharp
while (!eventProcessorClient.IsRunning && !cancellationToken.IsCancellationRequested)
{
    logger.LogInformation("Client status. IsRunning: {isRunning}", eventProcessorClient.IsRunning);
    await Task.Delay(TimeSpan.FromSeconds(10)).ConfigureAwait(false);
}
```

If there's a problem during start, I would expect to see it surface potentially a couple of different ways, depending on when in the flow it happens:

 - `StartAsync` would throw when awaited if there was an exception initializing state.  Once the main processing loop is initialized, it is run in the background, so the start call would lose visibility.  _(communicating with the service, takes place as part of the background, so credential issues would not appear here)_

- The error handler should be signaled if there were issues encountered while performing the main processing loop; you should see your handler fire for any issues communicating with the service, any load balancing issues, and any problems interacting with storage.

- `StopAsync` should surface any exceptions that caused the background processing to fault; this would also have been surfaced by the error handler, but awaiting the stop call would bubble the exception.

It may be helpful to see how the processor is being constructed and how a token is being acquired.   In addition, if you're willing, I'd like to try two things:

- Create an `EventHubConsumerClient` using the same set of parameters that you're passing to the processor constructor and then use that client to query the partitions using `EventHubConsumerClient::GetPartitionIdsAsync` and then try to read events using `EventHubConsumerClient::ReadEventsAsync`.   That would confirm that the basic parameters and authorization are allowing service communication and that there aren't any issues with the Event Hub itself.

- Add a bit of code to the following block:

```csharp
while (!eventProcessorClient.IsRunning && !cancellationToken.IsCancellationRequested)
{
    logger.LogInformation("Client status. IsRunning: {isRunning}", eventProcessorClient.IsRunning);
    await Task.Delay(TimeSpan.FromSeconds(10)).ConfigureAwait(false);
}
```

After your delay, I'd like to see what happens if we add:
```csharp
await eventProcessorClient.StopAsync().ConfigureAwait(false);
```

In the meantime, I'll try to replicate your sample and see if I'm able to reproduce locally.  Any additional context that you can think of around the configuration of your Event Hubs namespace, the Event Hub itself, how you're initializing the processor, or your general host environment would be very much appreciated.

**Comment by joebeernink** [comment]
Thanks @jsquire for the quick response. A couple of notes:

Yes, the IsRunning state never goes to tru, but also, the event handlers never fire, even though I'm pushing messages onto the event Hub.

I wasn't aware that auth issues wouldn't bubble up here. I had assumed that creation of the client would throw if there were problems authenticating. Here's the code I am using to get the token. I am using ManagedIdentityCredential... which could be the source of the issue. 

```
        public EventProcessorClient GetEventProcessorClient()
        {
            try
            {
                var eventHubConfiguration = dynamicConfigurationProvider.Get<EventHubConfiguration>();
                var apimLogToGenevaConfiguration = dynamicConfigurationProvider.Get<ApimLogToGenevaConfiguration>();

                logger.LogInformation("Getting Token Credential for Event Hub Namespace: {namespace}, Event Hub Name: {name}, Consumer Group: {consumerGroup}", eventHubConfiguration.Namespace, eventHubConfiguration.Name.ToLowerInvariant(), eventHubConfiguration.ConsumerGroup);

                var tokenCredential = new ManagedIdentityCredential();
                var storageAccountName = apimLogToGenevaConfiguration.LogStorageAccountName;
                var endpointSuffix = apimLogToGenevaConfiguration.AzureEndpointSuffix;
                var blobContainerUrl = new Uri($"https://{storageAccountName}.{endpointSuffix}/{leaseContainerName}");

                var blobContainerClient = new BlobContainerClient(blobContainerUrl, tokenCredential);
                return new EventProcessorClient(blobContainerClient, eventHubConfiguration.ConsumerGroup, eventHubConfiguration.Namespace, eventHubConfiguration.Name.ToLowerInvariant(), tokenCredential);
            }
            catch (Exception ex)
            {
                logger.LogError(ex, "Failed to get EventHostProcessorClient");
                throw;
            }
        }

```

I'll try your ideas on trying to isolate the auth issue.

**Comment by joebeernink** [comment]
I've found part of the issue (or rather your debugging steps found the issue).

I was passing in sb://<namespacename>.servicebus.windows.net instead of just <namespacename>.servicebus.windows.net. 

System.UriFormatException: Invalid URI: The hostname could not be parsed.
   at System.Uri.CreateThis(String uri, Boolean dontEscape, UriKind uriKind)
   at System.Uri..ctor(String uriString)
   at System.UriBuilder.get_Uri()

Changing that to remove the sb:// from the connection got the IsRunning state to true.

Now I get:
Error on Partition: (null), Error: Retry failed after 6 tries. (No such host is known.) (No such host is known.) (No such host is known.) (No such host is known.) (No such host is known.) (No such host is known.) 

Investigating that

**Comment by jsquire** [comment]
Hi Joe.    Thank you for the additional context.   The error seems appropriate given the input, but I'm concerned that the error handler didn't fire.   We'll definitely be poking at that.

The latest error also seems to make sense.   I'd speculate that we're at the point where we are trying to query the Event Hub partitions and unable to resolve the service address.

**Comment by jsquire** [comment]
Confirming that there is a bug in the processor initialization; we'll get that patched up for the next release.    Apologies for the difficulties.

**Comment by jsquire** [comment]
To bring this up to date,  Joe and I spoke offline and we managed to unblock his work.  

The initial issue that Joe reported is a previously unknown bug in the processor initialization in which a failure to create the `EventHubConnection` used when the processor is starting up is not in a fault-tolerant block and causes the processor background task to fail without being logged or surfaced to the error handler.

Marking the issue up and pulling into the April milestone for a fix.

**Comment by jsquire** [comment]
The root of this was addressed in #10622 where validation was included for well-formedness of the Event Hubs namespace.   #9324 will cover ensuring that the main processing loop does not perform any actions outside of a try/catch block so that any issues are surfaced via the error handler.

**Comment by shivan19** [comment]
I'm sending incorrect credentials as part of the connection string and using EventProcessorClient to start and stop processing. Apparently, it doesn't throw an exception stating 401 or Invalid credentials. How should I test the event hub connection and report bad credentials, in case?

**Comment by jsquire** [comment]
@shivan19: You should be observing this case with your error handler.  The service has responsibility for validating your credentials and making a decision on authorization.  The client will respond with an `UnauthorizedAccessException` if the service rejects a connection/link as unauthorized.   If your credentials are malformed,  it is possible that you'll to see a more generic exception because the server cannot parse the values and instead returns a non-specific server error.

If you're seeing something that you don't expect or would like to discuss further, please open an issue with the full context and any sample code/messages that you can provide.

**Comment by shivan19** [comment]
@jsquire : Opened a new issue #14244.
